### ISSUE TITLE
Performance: Improve attribute deletion to take advantage of indexes

### ISSUE THREAD

**Issue by nei** [magento/magento2#12804]
This issue is to discuss a alternative way to do deletion in order to use indexes and improve performance as well avoid locks on the table.

The table catalog_product_entity_int which I have about 6,6 million records and a delete statement for this table is taking up to 40 seconds to finish and seems to be locking some other operations on this table. Which is causing a huge slowness on the api and massive spikes on response time.

That is my NR graph on period where we updated our attribute sets definition.
![sql-attribute-deletion-spike](https://user-images.githubusercontent.com/33385/34156651-0b7ce59a-e4b6-11e7-95ac-a216be9b44c6.png)

The problem is that we have concurrency on delete update and selects on this table. 

Here is the execution plan for the deletion that is happening on core magento in here https://github.com/magento/magento2/blob/2.1-develop/app/code/Magento/Catalog/Model/ResourceModel/Attribute.php#L156

```
EXPLAIN  DELETE FROM `catalog_product_entity_int` 
WHERE 
    (attribute_id =99) AND 
    (row_id IN (
        (SELECT `catalog_product_entity`.`row_id` FROM `catalog_product_entity` WHERE (attribute_set_id = 25030) AND 
        (catalog_product_entity.created_in <= ' 1' ) AND (catalog_product_entity.updated_in > '2147483647'))
    )) ;
```

Explain shows that 6493539 rows being scanned:

```
    *************************** 1. row ***************************
           id: 1
  select_type: PRIMARY
        table: catalog_product_entity_int
         type: ALL
possible_keys: PRIMARY,CATALOG_PRODUCT_ENTITY_INT_ENTITY_ID_ATTRIBUTE_ID_STORE_ID,CATALOG_PRODUCT_ENTITY_INT_ATTRIBUTE_ID,CATALOG_PRODUCT_ENTITY_INT_STORE_ID
          key: NULL
      key_len: NULL
          ref: NULL
         rows: 6493539
        Extra: Using where
*************************** 2. row ***************************
           id: 2
  select_type: DEPENDENT SUBQUERY
        table: catalog_product_entity
         type: unique_subquery
possible_keys: PRIMARY,CATALOG_PRODUCT_ENTITY_ATTRIBUTE_SET_ID,CATALOG_PRODUCT_ENTITY_CREATED_IN,CATALOG_PRODUCT_ENTITY_UPDATED_IN
          key: PRIMARY
      key_len: 4
          ref: func
         rows: 1
        Extra: Using where
2 rows in set (0.00 sec)
```

I am not an Mysql expert but It seems Mysql will not use index for sub query on a range (IN) clause. From my research the only solution is to rewrite the query to do delete with JOIN.

Here is the rewrite for the delete.
```
EXPLAIN  DELETE a FROM `catalog_product_entity_int` AS a
        INNER JOIN `catalog_product_entity` AS b ON (a.row_id = b.row_id )
        WHERE 
         a.attribute_id = 99 AND 
         b.attribute_set_id = 25030 AND 
         b.created_in <= 1  AND 
         b.updated_in > 2147483647
```
As you can see the execution plan is much better.
```
*************************** 1. row ***************************
           id: 1
  select_type: SIMPLE
        table: b
         type: ref
possible_keys: PRIMARY,CATALOG_PRODUCT_ENTITY_ATTRIBUTE_SET_ID,CATALOG_PRODUCT_ENTITY_CREATED_IN,CATALOG_PRODUCT_ENTITY_UPDATED_IN
          key: CATALOG_PRODUCT_ENTITY_ATTRIBUTE_SET_ID
      key_len: 2
          ref: const
         rows: 1
        Extra: Using where
*************************** 2. row ***************************
           id: 1
  select_type: SIMPLE
        table: a
         type: ref
possible_keys: CATALOG_PRODUCT_ENTITY_INT_ENTITY_ID_ATTRIBUTE_ID_STORE_ID,CATALOG_PRODUCT_ENTITY_INT_ATTRIBUTE_ID
          key: CATALOG_PRODUCT_ENTITY_INT_ENTITY_ID_ATTRIBUTE_ID_STORE_ID
      key_len: 6
          ref: dtmagento.b.row_id,const
         rows: 1
        Extra: NULL
2 rows in set (0.00 sec)
```

What you guys think about this? 

### Preconditions
1. Magento 2.1.7
2. PHP 7.0.25
3. Aurora Database

### Steps to reproduce
1. Create hundreds of requests to API do delete/update your attribute definition

### Expected result
1. Requests complete successfully

### Actual result
1. Requests complete successfully but with a big impact on the response time and mysql load server.

**Comment by magento-engcom-team** [comment]
@nei, thank you for your report.
We've created internal ticket(s) MAGETWO-85917 to track progress on the issue.

**Comment by magento-engcom-team** [comment]
Hi @nei. Thank you for your report. 
The issue has been fixed in magento-engcom/magento2ce#1110 by @serhii-balko in 2.2-develop branch
Related commit(s):
 - [16575d92ae525e26607c187a805ebf00138c1612](https://github.com/magento/magento2/commit/16575d92ae525e26607c187a805ebf00138c1612)
 - [25fc7e85fbb942503b7f77de47a92f1d5917be36](https://github.com/magento/magento2/commit/25fc7e85fbb942503b7f77de47a92f1d5917be36)
 - [11da0ea0de76c729ff467ac67d29bcb04ababf59](https://github.com/magento/magento2/commit/11da0ea0de76c729ff467ac67d29bcb04ababf59)
 - [0b01926f5dfd37df48d162b3a00aed7f5aaf405a](https://github.com/magento/magento2/commit/0b01926f5dfd37df48d162b3a00aed7f5aaf405a)
 - [4fb932f2161f2e76ed10726375cab853160da135](https://github.com/magento/magento2/commit/4fb932f2161f2e76ed10726375cab853160da135)

The fix will be available with the upcoming 2.2.5 release.
